crypto/evp/kdf_lib.c

```diff
@@ -144,6 +145,138 @@ int EVP_KDF_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t keylen,
+    };
+
+    if (ctx == NULL)
+        return 0;
+
+    ckey.name = (paramname != NULL) ? paramname : OSSL_KDF_PARAM_KEY;
+
+    if (ctx->meth->set_skey != NULL && ctx->meth->prov == key->skeymgmt->prov)
+        return ctx->meth->set_skey(ctx->algctx, key->keydata, ckey.name);
+
+    /*
+     * We can't use the opaque key directly.
+     * Let's try to export it and set the ctx params in a traditional manner.
+     */
+    ckey.param = &params[0];
+
+    if (!ctx->meth->set_ctx_params)
+        return 0;
+
+    if (EVP_SKEY_export(key, OSSL_SKEYMGMT_SELECT_SECRET_KEY,
+                        convert_key_cb, &ckey))
+        return ctx->meth->set_ctx_params(ctx->algctx, params);
+
+    return 0;
+}
+
+EVP_SKEY *EVP_KDF_derive_SKEY(EVP_KDF_CTX *ctx, const char *key_type,
+                              size_t keylen, const char *propquery,
+                              const OSSL_PARAM params[])
+{
+    EVP_SKEYMGMT *skeymgmt = NULL;
+    EVP_SKEY *ret = NULL;
+
+    if (ctx == NULL || key_type == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+        return NULL;
+    }
+
+    skeymgmt = evp_skeymgmt_fetch_from_prov(ctx->meth->prov,
+                                            key_type, propquery);
+    if (skeymgmt == NULL) {
+        /*
+         * The provider does not support skeymgmt, let's try to fallback
+         * to a provider that supports it
+         */
+        skeymgmt = EVP_SKEYMGMT_fetch(ossl_provider_libctx(ctx->meth->prov),
+                                      key_type, propquery);
+    }
+
+    if (skeymgmt == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_FETCH_FAILED);
+        return NULL;
+    }
+
+    /* Fallback to raw derive + import if necessary */
+    if (skeymgmt->prov != ctx->meth->prov ||
+        ctx->meth->derive_skey == NULL) {
+        unsigned char *key = NULL;
+
+        if (ctx->meth->derive == NULL) {
+            ERR_raise(ERR_R_EVP_LIB, ERR_R_UNSUPPORTED);
+            return NULL;
+        }
+
+        key = OPENSSL_zalloc(keylen);
+        if (!key) {
+            EVP_SKEY_free(ret);
+            return NULL;
+        }
+
+        if (!ctx->meth->derive(ctx->algctx, key, keylen, params)) {
+            EVP_SKEY_free(ret);
+            OPENSSL_free(key);
+            return NULL;
+        }
+
+        ret = EVP_SKEY_import_raw_key(ossl_provider_libctx(skeymgmt->prov),
+                                      key_type, key, keylen, propquery);
+        EVP_SKEYMGMT_free(skeymgmt);
+
+        OPENSSL_clear_free(key, keylen);
+        return ret;
+    }
+
+    ret = evp_skey_alloc(skeymgmt);
+    if (ret == NULL) {
+        EVP_SKEYMGMT_free(skeymgmt);
+        return NULL;
+    }
+
+    ret->skeymgmt = skeymgmt;
+
+    ret->keydata = ctx->meth->derive_skey(ctx->algctx, ossl_provider_ctx(skeymgmt->prov),
+                                          skeymgmt->import, keylen, params);
+    if (ret->keydata == NULL) {
+        EVP_SKEY_free(ret);
+        EVP_SKEYMGMT_free(skeymgmt);
```

Comment:
Isn't this a double free? We already assigned `skeymgmt` to `ret` - so isn't this freed via `EVP_SKEY_free`. I'm wondering why this doesn't get detected by the tests.